test/base-fork/Morpho.t.sol

```diff
@@ -0,0 +1,272 @@
+        );
+        // USDC idle market already exists
+        IMetaMorpho(MORPHO_VAULT_USDS).submitCap(
+            usdsParams,
+            type(uint184).max
+        );
+        IMetaMorpho(MORPHO_VAULT_USDC).submitCap(
+            usdcParams,
+            type(uint184).max
+        );
+
+        skip(1 days);
+
+        IMetaMorpho(MORPHO_VAULT_USDS).acceptCap(usdsParams);
+        IMetaMorpho(MORPHO_VAULT_USDC).acceptCap(usdcParams);
+        
+        Id[] memory supplyQueueUSDS = new Id[](1);
+        supplyQueueUSDS[0] = MarketParamsLib.id(usdsParams);
+        IMetaMorpho(MORPHO_VAULT_USDS).setSupplyQueue(supplyQueueUSDS);
+        Id[] memory supplyQueueUSDC = new Id[](1);
+        supplyQueueUSDC[0] = MarketParamsLib.id(usdcParams);
+        IMetaMorpho(MORPHO_VAULT_USDC).setSupplyQueue(supplyQueueUSDC);
+
+        rateLimits.setRateLimitData(
+            RateLimitHelpers.makeAssetKey(
+                foreignController.LIMIT_VAULT_DEPOSIT(),
+                MORPHO_VAULT_USDS
+            ),
+            25_000_000e18,
+            uint256(5_000_000e18) / 1 days
+        );
+        rateLimits.setRateLimitData(
+            RateLimitHelpers.makeAssetKey(
+                foreignController.LIMIT_VAULT_DEPOSIT(),
+                MORPHO_VAULT_USDC
+            ),
+            25_000_000e6,
+            uint256(5_000_000e6) / 1 days
+        );
+
+        vm.stopPrank();
+    }
+
+    function _getBlock() internal pure override returns (uint256) {
+        return 22841965;  // November 24, 2024
+    }
+
+}
+
+/**********************************************************************************************/
+/*** Only testing USDS failure modes because it is the same for USDC                        ***/
+/**********************************************************************************************/
+
+contract MorphoFailureTests is MorphoBaseTest {
+
+    function test_morpho_deposit_notRelayer() external {
+        vm.expectRevert(abi.encodeWithSignature(
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_deposit_frozen() external {
+        vm.prank(freezer);
+        foreignController.freeze();
+
+        vm.prank(relayer);
+        vm.expectRevert("ForeignController/not-active");
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_deposit_rateLimited() external {
+        deal(Base.USDS, address(almProxy), 25_000_001e18);
+
+        assertEq(usdsVault.convertToAssets(usdsVault.balanceOf(address(almProxy))), 0);
+        assertEq(IERC20(Base.USDS).balanceOf(address(almProxy)),                    25_000_001e18);
+
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        vm.prank(relayer);
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_001e18);
```

Comment:
`25_000_000e18 + 1`, applies to all boundary conditions

Reply:
updated